File: services/todo/todo.py
#!/usr/bin/python3
import functools
import json
import logging
import os
import typing

import flask
import opencensus.ext.flask.flask_middleware
import opencensus.ext.sqlalchemy.trace
import opencensus.ext.zipkin.trace_exporter
import opencensus.trace.samplers
import opencensus.trace.tracer
import sqlalchemy as sa
import sqlalchemy.ext.declarative
import sqlalchemy.orm

app = flask.Flask(__name__)

db_url = 'postgresql://{usr}:{pswd}@database:5432/{db}'.format(
    usr=os.getenv("POSTGRES_USER"),
    pswd=os.getenv("POSTGRES_PASSWORD"),
    db=os.getenv("POSTGRES_DB")
)
engine = sa.create_engine(db_url)
logger = logging.getLogger()

Session: typing.Type[sa.orm.Session] = typing.cast(typing.Type[sa.orm.Session], sa.orm.sessionmaker(bind=engine))
Base = sa.ext.declarative.declarative_base()

trace_exporter = opencensus.ext.zipkin.trace_exporter.ZipkinExporter(
    service_name="Todo",
    host_name="zipkin.istio-system.svc.cluster.local",
    port=9411,
)
trace_sampler = opencensus.trace.samplers.ProbabilitySampler(1)
tracer = opencensus.trace.tracer.Tracer(
    sampler=trace_sampler,
    exporter=trace_exporter,
)
opencensus.ext.sqlalchemy.trace.trace_integration(tracer)
opencensus.ext.flask.flask_middleware.FlaskMiddleware(app, sampler=trace_sampler, exporter=trace_exporter)


class List(Base):
    __tablename__ = 'list'
    id = sa.Column(sa.Integer, primary_key=True)

    name = sa.Column(sa.String)

    items = sa.orm.relationship('Item')

    def to_dict(self) -> dict:
        return {
            'id': self.id,
            'name': self.name,
            'items': self.items
        }


class Item(Base):
    __tablename__ = 'item'
    id = sa.Column(sa.Integer, primary_key=True)

    fk_list = sa.Column(sa.Integer, sa.ForeignKey('list.id'), nullable=False)

    content = sa.Column(sa.String)

    def to_dict(self) -> dict:
        return {
            'id': self.id,
            'name': self.content
        }


def debug(fn):
    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        if flask.request.data:
            logger.error("Handling incoming request: {}".format(json.loads(flask.request.data)))
        result: flask.Response = fn(*args, **kwargs)
        logger.error("Returning: {}".format(result))
        return result

    return wrapper


class SessionContext:
    def __init__(self):
        self._session = Session()

    def __enter__(self):
        return self._session

    def __exit__(self, exc_type, exc_val, exc_tb):
        try:
            self._session.commit()
        finally:
            self._session.close()


def load_lists(session) -> flask.Response:
    response: flask.Response = flask.jsonify([{
        'id': list_.id,
        'name': list_.name,
        'items': [
            {
                'id': item.id,
                'content': item.content,
                'fk_list': item.fk_list
            }
            for item in list_.items
        ],
    } for list_ in session.query(List)])
    response.status_code = 201
    return response


@app.route('/list', methods=['POST'])
@debug
def create_list():
    data = json.loads(flask.request.data)
    if 'name' not in data:
        return "Missing required 'name' from input JSON data", 400

    with SessionContext() as session:
        todo_list = List(name=data['name'])
        session.add(todo_list)
        return load_lists(session)


@app.route('/list/<lid>/item', methods=['POST'])
@debug
def create_item(lid: int):
    data = json.loads(flask.request.data)
    if 'content' not in data:
        return "Missing required 'content' from input JSON data", 400

    with SessionContext() as session:
        list_ = session.query(List).get(lid)
        list_.items.append(Item(content=data['content']))
        return load_lists(session)


@app.route('/list', methods=['GET'])
@debug
def get_lists():
    with SessionContext() as session:
        return load_lists(session)


if __name__ == '__main__':
    Base.metadata.create_all(engine)
    app.run("localhost", "9080")

File: services/todo/todo.go
package main

import (
	"encoding/json"
	"github.com/go-pg/pg/v9"
	"github.com/go-pg/pg/v9/orm"
	"log"
	"net/http"
	"os"
)

type Item struct {
	Id      int32
	Content string
}

type List struct {
	Id    int64
	Name  string
	Items []*Item
}

func getDB() *pg.DB {
	return pg.Connect(&pg.Options{
		Addr:     "database:5432",
		User:     os.Getenv("POSTGRES_USER"),
		Password: os.Getenv("POSTGRES_PASSWORD"),
		Database: os.Getenv("POSTGRES_DB"),
	})
}

func writeResponse(db *pg.DB, w http.ResponseWriter) {
	var lists []List

	err := db.Model(&lists).Select()
	if err != nil {
		http.Error(w, "Failed to marshal response", http.StatusInternalServerError)
	}

	response, err := json.Marshal(lists)
	if err != nil {
		http.Error(w, "Failed to marshal response", http.StatusInternalServerError)
	}

	w.Header().Set("Content-Type", "application/json")

	_, err = w.Write(response)
	if err != nil {
		http.Error(w, "Failed to write response", http.StatusInternalServerError)
	}
}

func listHandler(w http.ResponseWriter, r *http.Request) {
	db := getDB()
	defer db.Close()

	if r.Method == "POST" {
		var list List

		err := json.NewDecoder(r.Body).Decode(&list)
		if err != nil {
			http.Error(w, "Failed to decode json", http.StatusInternalServerError)
		}

		err = db.Insert(list)
		if err != nil {
			http.Error(w, "Failed to create new list", http.StatusInternalServerError)
		}

		writeResponse(db, w)
	} else if r.Method == "GET" {
		writeResponse(db, w)
	} else {
		http.Error(w, "Unknown method", http.StatusBadRequest)
	}
}

func createItem(w http.ResponseWriter, r *http.Request) {
	db := getDB()
	defer db.Close()

	if r.Method == "POST" {
		var item Item

		err := json.NewDecoder(r.Body).Decode(&item)
		if err != nil {
			http.Error(w, "Failed to decode json", http.StatusInternalServerError)
		}

		err = db.Insert(item)
		if err != nil {
			http.Error(w, "Failed to create new item", http.StatusInternalServerError)
		}

		writeResponse(db, w)
	} else {
		http.Error(w, "Unknown method", http.StatusBadRequest)
	}
}

func createSchema(db *pg.DB) error {
	for _, model := range []interface{}{(*Item)(nil), (*List)(nil)} {
		err := db.CreateTable(model, &orm.CreateTableOptions{
			Temp: true,
		})
		if err != nil {
			return err
		}
	}
	return nil
}

func main() {
	db := getDB()
	defer db.Close()

	err := createSchema(db)
	if err != nil {
		panic(err)
	}

	http.HandleFunc("/list", listHandler)
	http.HandleFunc("/list/:lid/item", createItem)

	log.Fatal(http.ListenAndServe(":9080", nil))
}
